### DISCUSSION TITLE
Use JSON-LD for structured data in storefront

### DISCUSSION THREAD

**Discussion by aragon999** [shopware/shopware#3170] [RFC / Proposal]
**Effort**: middle/high

### Description: 
Currently it is quite annoying to modify or add structured data to a page, as the information is distributed inline over multiple TWIG templates. Also for several places there are missing blocks, and it would require quite an amount of blocks to make every structured data editable, which is also quite an overhead.

Furthermore when reusing blocks in other places, there might be some unwanted structured data contained in the templates.

Therefore a better solution might be to implement the structured data using [JSON-LD](https://json-ld.org/). Maybe an idea would be to implement it similarly, as the `MetaInformation` for each page: https://github.com/shopware/platform/blob/6fe1f2a7d7722eb6ad6e8b6008561e73c9828248/src/Storefront/Page/Page.php#L35-L38

I am not sure how one can validate the correctness of such structured data, and what the best possible data structure is, but there currently exist two PHP libraries to work with such data:
https://github.com/lanthaler/JsonLD
https://github.com/digitalbazaar/php-json-ld

### Benefits
Easier extensibility of structured data.

### Break strategy
I think this might be quite challenging, to implement such a mechanism in a backwards compatible way, so the best way might be to start with 6.6 to completely switch to the new implementation.

Take for example the `buy_widget_data_rich_snippet_price_range` block:
https://github.com/shopware/platform/blob/6fe1f2a7d7722eb6ad6e8b6008561e73c9828248/src/Storefront/Resources/views/storefront/component/buy-widget/buy-widget.html.twig#L70-L88

Here the complete content depends on the available TWIG variables, if one would for example use something like:
```twig
{% block buy_widget_data_rich_snippet_price_range %}
    {% with {product: {calculatedPrices: [...]} %}
        {{ parent() }}
    {% endwith %}
{% endblock %}
```
One can completely modify the output of this metadata. If one would use the variables of the new data structure directly, the changes from the template would not be considered.

**Comment by JoshuaBehrens** [comment]
I already thought about it to have a JSON renderer for it as it is just a different representation of the already existing data. This moves the complexity of adding SEO data from someone, that understands HTML to someone who knows how to change a JS object. You will still likely have multiple snippets per page as the content is super dynamic. An advantage to pull this out of twig is the possible reusage in different storefront implementations

[a comment by an automated account is not shown]

**Comment by Phil23** [comment]
Hey folks, didn't know this discussion existed. I just pulled of a first iteration which replaces the inline data with JSON-LD. For now I just created separate templates that could be extended. Later we could improve this further with editing in the Admin and corresponding API routes, etc.

https://github.com/shopware/shopware/pull/15541

**Comment by keulinho** [comment]
As the feature has already landed in trunk (behind feature flag though for now) i'm closing this discussion for now. 
If you have follow up feedback, ideas for improvements or see issues in the current implementation, please create a regular issue for that.

Thanks for bringing this up! 🚀
